### ISSUE TITLE
Hide collision shapes when "visible" property is false

### ISSUE THREAD

**Issue by trollusk** [godotengine/godot-proposals#6857]
### Describe the project you are working on

Procedural city generation.

### Describe the problem or limitation you are having in your project

When debugging a scene it's often useful to hide branches of the scene tree (closed eye icon), to narrow down where the problem is. When "Debug > Show Collision Shapes" is true, collision shapes are never hidden in the playing scene, regardless of whether their node is visible or invisible. To debug collision shapes one has to go through the scenetree and manually click "disable" in the inspector for each individual collision shape/polygon node.

This is inconsistent with the behaviour of the SceneTree panel in the edited scene. There, hiding a collision shape node will make its shape invisible.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Collision shapes should either become invisible or display in a different colour when their node is hidden.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The behaviour of "Show Collision Shapes" would be changed so that collision shapes belonging to hidden nodes are not displayed, or are shown in a different colour or style (eg gray, with dotted lines).

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Not really, but I imagine it would be fairly easy to implement in the engine.

### Is there a reason why this should be core and not an add-on in the asset library?

This is a basic improvement to the usability of the "Show Collision Shapes" debug facility, to make it more useful as well as consistent with the editor's treatment of collision shape visibility.

**Comment by Calinou** [comment]
Invisible collision shapes are still effective, so I think they should remain visible. Instead, having the ability to [customize colors on a per-shape basis](https://github.com/godotengine/godot-proposals/issues/906) will allow you to hide certain shapes while keeping others visible.

**Comment by KoBeWi** [comment]
But the shapes **are** disappearing when you make them invisible at runtime. Not sure what the proposal is about, can you show an example?

**Comment by trollusk** [comment]
Here is a running scene containing a StaticBody3D with a cube mesh, and a sphere CollisionShape3D.
![image](https://github.com/godotengine/godot-proposals/assets/220468/9d7694f3-95df-456b-9a79-90fa3c58856f)

Here is what you see when you hide every node in the scene:
![image](https://github.com/godotengine/godot-proposals/assets/220468/5b5cdf8a-ede3-4b55-bc48-12bcc7787b4c)

**Comment by KoBeWi** [comment]
I'd say this is a bug, because the shapes disappear correctly in 2D.

**Comment by Zireael07** [comment]
Yep, I even reported shapes disappearing as a proposal or a bug report once as I wanted them to stay :P

Different strokes, different folks...

**Comment by KoBeWi** [comment]
I looked into this and the 3D collision shapes rendering implementation is backwards. I was confused why my shape was invisible when I tested this initially, and it turns out that shapes don't draw themselves, but the parent collision object draws them. At least at runtime; it's different in the editor, where the shape appears as a gizmo.

The proposal makes sense, but it's not trivial to implement .-.

**Comment by trollodel** [comment]
The debug collision shape rendering for 3D changes in godotengine/godot#45783 (which describes why it's useful) and follow-up PRs. The problem described here was introduced in godotengine/godot#48175
Like @KoBeWi says, CollisionShape3D can't hide the debug shape. But the CollisionObject3D host can, so it's possible to partially restore the old behavior, to be closer to 2D. I'll do a PR soon for that.

>  At least at runtime; it's different in the editor, where the shape appears as a gizmo.

CollisionObject3D gizmo shapes can't be edited, so the ability to show them was left to CollisionShape3D and CollisionPolygon3D.

Also, it would be nice to have a 2D version of godotengine/godot#45783, if it's not already done.

**Comment by MajorMcDoom** [comment]
I think this is a band-aid solution for a very specific quality-of-life improvement.

Visibility is a core property which is albeit kind of useless for collision shapes. However, co-opting it for fine-tuning a debug visualization feature does not set a good precedent and will lead to future inconsistencies and more bandaid solutions.

For example, CSG shapes also show up in debug collision visualization. However, CSG shapes also have a visible component in addition to physical. This means invisible yet collidable CSGs are a legitimate possibility. The "rule" introduced by this PR cannot apply in the same way to CSGs, because the visibility is actually meaningful in that case. This leads to an inconsistency.

In short, `Visible` is core, and has nothing to do with collision debug visibility. I think introducing different colors to debug collision wires or just simply adding a distance fade would be a better way to reduce the visual clutter.

**Comment by natanmaia95** [comment]
This issue still exists as of Godot 4.5.

Personally, I want some way to know which of my Area3Ds (hitboxes, hurtboxes) are "enabled" or not when I control their monitoring/monitorable properties, regardless of their internal collision shapes, for monitoring active attack frames, or just hiding them for less clutter.

My current alternative is looping through all children collision shapes and changing their material, which feels clunky. I also often use fake debug collisions from CSG nodes like boxes and capsules, not ideal but simple enough, so a proposal should be at least as simple as that.

For proposals, I'd personally be satisfied if below debug color there was a toggle to hide debug draw when not visible, or draw with less transparency/different border style. I don't know if my specific use of hiding when itself or the parent Area is removed from processing is doable or flexible, and it's very useful to see collisions from hidden objects that are still solid, like invisible walls.
